### ISSUE TITLE
PyUpgrade Rules: Respect `target-version`

### ISSUE THREAD

**Issue by smsegal** [astral-sh/ruff#5035]
We need to set our `target-version` to `py37` in our codebase. When enabling the `UP` ruleset, `ruff` will use `pyupgrade` rules that are invalid for python 3.7. 

An example: 

```toml
# pyproject.toml snippet
[tool.ruff]
target-version = "py37"
select = ["UP"]
```

```python3
from typing import Tuple

# ruff reports "UP006: Use tuple instead of Tuple"
def foo(x: int, y: int) -> Tuple[int, int]: 
    return (x, y)
```

This is invalid for python versions below 3.9 (see the deprecation notice at https://docs.python.org/3.11/library/typing.html#typing.Tuple)

Details: 

Using `ruff 0.0.272` on MacOS, but reproduces on linux hosts as well. 

Command to reproduce using the above python snippet: 

```bash
ruff check --isolated --select "UP" --target-version "py37" foo.py
```

**Comment by charliermarsh** [comment]
Hmm, I ran that exact command (`ruff check --isolated --select "UP" --target-version "py37" foo.py -n`), on that exact snippet, and didn't get any `UP006` error.

Do you perhaps have `from __future__ import annotations` in that file? Or is it a `.pyi` file?

**Comment by smsegal** [comment]
Ah, good catch! I didn't notice, but we do indeed have `__future__` annotations in that file. 
Curious, why do the `__future__` annotations cause this?

**Comment by smsegal** [comment]
Thanks for the info! That makes total sense. Are you aware of any downsides of importing `__future__.annotations`? Reading more about it, I see the PEP introducing it was delayed twice "to avoid breaking downstream code" but reading the announcements of the delays doesn't give any concrete details about what kind of breakages I should be worried about. 


(Also, really appreciate the fast turnaround time on issues in this repo. Excited to see what else comes from astral!)

**Comment by charliermarsh** [comment]
Totally!

The main downside, AFAIK, is that it can cause problems for libraries that _rely_ on accessing type annotations at runtime, like Pydantic and FastAPI. I think those libraries do support `__future__` annotations now, but there are still some quirks. For example, this will give you a runtime error if you're on < Python 3.9 (the `int | None` syntax introduced in Python 3.9):

```py
from __future__ import annotations

from pydantic import BaseModel

class Model(BaseModel):
    a: int | None

print(Model(a=1))
```

`a: int | None` is fine from Python's perspective in general , since you have `from __future__ import annotations`; but Pydantic has to resolve the type annotation to something useful to make it enforceable at runtime, and when it tries, you get a `TypeError`.

So, in short, if you use `__future__` annotations with those libraries, some annotations that Python can _parse_ without issue may explode at runtime.
